src/mlpack/core/util/params.hpp

```diff
@@ -0,0 +1,145 @@
+         const std::string& bindingName,
+         const BindingDetails& doc);
+
+  /**
+   * Return `true` if the specified parameter was given.
+   *
+   * @param identifier The name of the parameter in question.
+   */
+  bool Has(const std::string& identifier) const;
+
+  /**
+   * Get the value of type T found for the parameter specified by `identifier`.
+   * You can set the value using this reference safely.
+   *
+   * @param identifier The name of the parameter in question.
+   */
+  template<typename T>
+  T& Get(const std::string& identifier);
+
+  /**
+   * Cast the given parameter of the given type to a short, printable
+   * `std::string`, for use in status messages.  The message returned here
+   * should be only a handful of characters, and certainly no longer than one
+   * line.
+   *
+   * @param identifier The name of the parameter in question.
+   */
+  template<typename T>
+  std::string GetPrintable(const std::string& identifier);
+
+  /**
+   * Get the raw value of the parameter before any processing that Get() might
+   * normally do.  So, e.g., for command-line programs, this does not
+   * perform any data loading or manipulation like Get() does.  So if you
+   * want to access a matrix or model (or similar) parameter before it is
+   * loaded, this is the method to use.
+   *
+   * @param identifier The name of the parameter in question.
+   */
+  template<typename T>
+  T& GetRaw(const std::string& identifier);
+
+  /**
+   * Given two (matrix) parameters, ensure that the first is an in-place copy of
+   * the second.  This will generally do nothing (as the bindings already do
+   * this automatically), except for command-line bindings, where we need to
+   * ensure that the output filename is the same as the input filename.
+   *
+   * @param outputParamName Name of output (matrix) parameter.
+   * @param inputParamName Name of input (matrix) parameter.
+   */
+  // TODO: it would be really nice to remove this!  It's only used by MeanShift
+  // and KMeans bindings.
+  void MakeInPlaceCopy(const std::string& outputParamName,
+                       const std::string& inputParamName);
+
+  //! Get the map of parameters.
+  std::map<std::string, ParamData>& Parameters() { return parameters; }
+  //! Get the map of aliases.
+  std::map<char, std::string>& Aliases() { return aliases; }
+
+  //! Get the binding name.
+  const std::string& BindingName() const { return bindingName; }
+
+  //! Get the binding details.
+  const BindingDetails& Doc() const { return doc; }
+
+  /**
+   * Set the particular parameter as passed.
+   *
+   * @param identifier The name of the parameter to set as passed.
+   */
+  void SetPassed(const std::string& identifier);
+
+  /**
+   * Check all input matrices for NaN and inf values, and throw an exception if
+   * any are found.
+   */
+  void CheckInputMatrices();
+
+ private:
+  //! Convenience map from alias values to names.
+  std::map<char, std::string> aliases;
```

Comment:
I am not sure yet, but I think this functionality is specific to the command-line bindings and likely could be removed from this class, simplifying the code somewhat.  I'll keep looking into this in the upcoming days.

Reply:
It will simplify the code, but according to me using aliases through the command line is a lot easier instead of writing complete parameter names.

Reply:
Absolutely, I agree---but when looking through the code, it seems like CLI11 handles all the aliases and maps them back to the original parameter names.  So we may need to keep the `alias` field in `util::ParamData`, but I don't think we need this map in `util::Params`.  In fact, the `aliases` map here is only used in `Has()` and `Get()`, so that you could define a parameter `training` with alias `t`, and then call `Has("t")` in the binding.  I don't think we access parameters by their alias anywhere in the existing bindings, so we can probably remove this member.  But, if I do it, I'll do it in a follow-up PR.